### PR TITLE
Implement custom DNS resolver

### PR DESCRIPTION
**Pull request by dani-garcia**
The goal of this change is to protect us more thoroughly against DNS rebinding attacks and redirects in the icons service, previously we did a manual lookup check before doing the initial HTTP request, which would leave us vulnerable, by inserting a middleware DNS resolver into reqwest we should be protected against all cases.

Also I noticed that with the `hickory-resolver` crate we can enable DNS over TLS and DNS over HTTPS, which seems like a great option, but we'd need to test further.

Also moved the `is_global_ip` functions to utils, because it was crowding the already big `icons` file.

### CODE REVIEW

**Review by williamdes — APPROVED** [review]

**Review by BlackDex — COMMENTED** [review]
Other then the missing `cached` proc-macro it looks ok too me.
I tested it with DDoSing the Favicon endpoint, that seems to work ok.

I also did a longer test which test around 300 domains (including some duplicates with/without redirects). These all seem to work as on the current code.

File: src/api/icons.rs [hunk 1]
```diff
@@ -166,153 +147,30 @@ fn is_valid_domain(domain: &str) -> bool {
...
-}
-
-#[derive(Clone)]
-enum DomainBlacklistReason {
-    Regex,
-    IP,
-}
-
-use cached::proc_macro::cached;
-#[cached(key = "String", convert = r#"{ domain.to_string() }"#, size = 16, time = 60)]
```
[the remark this answers is not in the document]
Comment (dani-garcia) [reply src/api/icons.rs:275]: Well, initially I wanted to add the `cached` to here: https://github.com/dani-garcia/vaultwarden/blob/144925a1f55ce4511cf6dad8a74233098a833acc/src/util.rs#L836-L848

But cached doesn't work in functions to take self, and the Hickory resolver (previously named trust-dns) has internal caching itself, so it seemed like less of a problem.

We should definitely benchmark it anyway.

[the remark this answers is not in the document]
Comment (BlackDex) [reply src/api/icons.rs:275]: The mean reason for it being there was not the DNS lookup, but the regex match. That can take a lot of cpu cycles and causes a delay. No need to keep regex testing for a short period of time if it is in the cache.

The internal cache of hickory is good for all other items, but will not prevent the regex from getting triggered every time the same domain gets called.

**Comment by BlackDex** [comment]
Did some further testing and seems to work fine.
I only wonder if we do not want to have this custom resolver as a default for everything.
It can also be useful for the calls to HIBP, Bitwarden Push, Calls to DUO or YubiKey OTP.

Other than that, it's all good.

**Comment by PKizzle** [comment]
I just had a brief look at the PR so maybe I missed it but is it possible to disable this feature in the configuration? I am already using a custom secure DNS solution and rather use that one as it gives me a more fine-grain control.

**Comment by BlackDex** [comment]
> I just had a brief look at the PR so maybe I missed it but is it possible to disable this feature in the configuration? I am already using a custom secure DNS solution and rather use that one as it gives me a more fine-grain control.

It still uses the DNS servers provided by the host, so no need i think.

**Review by BlackDex — APPROVED** [review]

**Review by BlackDex — APPROVED** [review]
